src/util/string.h

```diff
@@ -17,6 +18,55 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <unsigned num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
```

Comment:
While it gets the job done, I think the parser is unnecessarily complicated for several reasons:
* Redundant Counters: Separate (but mutually exclusive) `count_normal` and `count_pos`.
* Repetition: The check for "Format specifier incorrectly terminated by end of string" is repeated in multiple places.
* Inefficient Character Handling: Multiple character checks for `%` (and manual digit detection).
* Scattered Iterations: Increment operations (++it), continue statements, and lookaheads (str.end()) make the control flow harder to follow.

I have a hard time explaining the behavior of the parser in words, following the logic of the code. Can we use a state-machine based parser instead, without lookaround, iterating char by char without jumping around, changing the state as we go along, e.g. something like this:
```C++
constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
{
    unsigned count{0};
    unsigned maybe_num{0};
    bool inside_format_specifier = false;
    bool has_normal = false, has_positional = false;
    for (auto it = str.begin(); it < str.end(); ++it) {
        if (*it == '%') inside_format_specifier = !inside_format_specifier;
        else if (inside_format_specifier) {
            if (std::isdigit(*it)) maybe_num = maybe_num * 10 + (*it - '0');
            else if (*it == '$') {
                if (maybe_num == 0) throw "Positional format specifier must be at least 1";
                count = std::max(count, maybe_num);
                has_positional = true;
                inside_format_specifier = false;
            } else {
                ++count;
                maybe_num = 0;
                has_normal = true;
                inside_format_specifier = false;
            }
        } else maybe_num = 0;
    }

    if (has_normal && has_positional) throw "Format specifiers must be all positional or all non-positional!";
    if (inside_format_specifier || maybe_num > 0) throw "Format specifier incorrectly terminated by end of string";
    if (num_params != count) throw "Format specifier count must match the argument count!";
}
```

This is the reason I'm not ack-ig yet.

Reply:
Thanks, updated the example